src/script/conversation/ConversationRepository.ts

```diff
@@ -179,85 +173,66 @@ export class ConversationRepository {
-            await this.addMissingMember(conversation, missingUsers, new Date(time).getTime() - 1);
-          }
-        }
+    this.messageRepository.setClientMismatchHandler(async (mismatch, conversationId, silent, consentType) => {
+      const conversation = conversationId ? await this.getConversationById(conversationId) : undefined;
+      const {missingClients, deletedClients, emptyUsers, missingUserIds} = extractClientDiff(
+        mismatch,
+        conversation?.allUserEntities,
+      );
 
-        const removedTeamUserIds = (
-          await Promise.all(
-            deletedClients.map(
-              ({userId, data}) => data.map(client => this.userRepository.removeClientFromUser(userId, client))[0],
-            ),
-          )
-        )
-          .filter(user => user?.devices().length === 0)
-          .filter(user => !!user)
-          .filter(user => user.inTeam())
-          .map(user => user.id);
-
-        if (removedTeamUserIds.length) {
-          // If we have found some users that were removed from the conversation, we need to check if those users were also completely removed from the team
-          const usersWithoutClients = await this.userRepository.getUserListFromBackend(removedTeamUserIds);
-          await Promise.all(
-            usersWithoutClients
-              .filter(user => user.deleted)
-              .map(user =>
-                this.teamMemberLeave(
-                  this.teamState.team().id,
-                  {
-                    domain: this.teamState.teamDomain(),
-                    id: user.id,
-                  },
-                  new Date(time).getTime() - 1,
-                ),
+      if (conversation && missingUserIds.length) {
+        // add/remove users from the conversation (if any)
+        await this.addMissingMember(conversation, missingUserIds, new Date(mismatch.time).getTime() - 1);
+      }
+
+      // Remove clients that are not needed anymore
+      deletedClients.forEach(({userId, clients}) =>
+        clients.forEach(client => this.userRepository.removeClientFromUser(userId, client)),
+      );
+      const removedTeamUserIds = emptyUsers.filter(user => user.inTeam()).map(user => user.qualifiedId);
+
+      if (removedTeamUserIds.length) {
+        // If we have found some users that were removed from the conversation, we need to check if those users were also completely removed from the team
+        const usersWithoutClients = await this.userRepository.getUserListFromBackend(
+          conversation.isFederated() ? removedTeamUserIds : removedTeamUserIds.map(({id}) => id),
+        );
+        await Promise.all(
+          usersWithoutClients
+            .filter(user => user.deleted)
+            .map(user =>
+              this.teamMemberLeave(
+                this.teamState.team().id,
+                {
+                  domain: this.teamState.teamDomain(),
+                  id: user.id,
+                },
+                new Date(mismatch.time).getTime() - 1,
```

Comment:
Just for curiosity.. why we use `.getTime() - 1` in some places in this file?

Reply:
This is in order to inject the message before the message being sent. 
This is a not so nice hack to have message ordered the right way. 

In a not so distant future, I'd like to have a message buffer that takes care of putting the right timestamp to system messages, but for now we are stuck with this system

Reply:
got it thanks for explanation